indexer/editable_map_object.cpp

```diff
@@ -213,4 +185,134 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
+  return strings::to_uint64(buildingLevels, levels) && levels <= kMaximumLevelsEditableByUsers;
+}
+
+// static
+bool EditableMapObject::ValidateHouseNumber(string const & houseNumber)
+{
+  if (houseNumber.empty())
+    return true;
+
+  strings::UniString us = strings::MakeUniString(houseNumber);
+  // TODO: Improve this basic limit - it was choosen by @Zverik.
+  auto constexpr kMaxHouseNumberLength = 15;
+  if (us.size() > kMaxHouseNumberLength)
+    return false;
+
+  // TODO: Should we allow arabic numbers like U+0661 ١	Arabic-Indic Digit One?
+  strings::NormalizeDigits(us);
+  for (strings::UniChar const c : us)
+  {
+    // Valid house numbers contain at least one number.
+    if (c >= '0' && c <= '9')
+      return true;
+  }
+  return false;
+}
+
+// static
+bool EditableMapObject::ValidateFlats(string const & flats)
+{
+  auto flatRangeIterator = strings::SimpleTokenizer(flats, ",;");
+  for (; flatRangeIterator != strings::SimpleTokenizer(); ++flatRangeIterator)
+  {
+    auto flatRange = *flatRangeIterator;
+    strings::Trim(flatRange);
+    vector<string> range(strings::SimpleTokenizer(flatRange, "-"), strings::SimpleTokenizer());
+    if (range.size() > 2)
+      return false;
+
+    for (auto const & rangeBorder : range)
+    {
+      if (!all_of(begin(rangeBorder), end(rangeBorder), isalnum))
+        return false;
+    }
+  }
+  return true;
+}
+
+// static
+bool EditableMapObject::ValidatePostCode(string const & postCode)
+{
+  return search::v2::LooksLikePostcode(postCode, false /* check whole matching */);
+}
+
+// static
+bool EditableMapObject::ValidatePhone(string const & phone)
+{
+  auto start = begin(phone);
+  auto const stop = end(phone);
+
+  auto const startsWithPlus = *start == '+';
+
+  auto const maxNumberLen = 12;
+  auto const minNumberLen = 7;
+
+  if (startsWithPlus)
+    ++start;
+
+  auto digitsCount = 0;
+  for (; start != stop; ++start)
+  {
+    auto const isCharValid = isdigit(*start) || *start == '(' ||
+                             *start == ')' || *start == ' ' || *start == '-';
+    if (isdigit(*start))
+      ++digitsCount;
+    if (!isCharValid)
+      return false;
+  }
+
+  return minNumberLen <= digitsCount && digitsCount <= maxNumberLen;
+}
+
+// static
+bool EditableMapObject::ValidateFax(string const & fax)
+{
+  return ValidatePhone(fax);
+}
+
+// static
+bool EditableMapObject::ValidateWebsite(string const & site)
+{
+  if (site.empty())
+    return true;
+
+  auto const dotPos = find(begin(site), end(site), '.');
+  // site should contain at least one dot but not at the begining and no at the and.
+  if (dotPos == end(site) || site.front() == '.' || site.back() == '.')
+    return false;
+
+  return true;
+}
+
+// static
+bool EditableMapObject::ValidateEmail(string const & email)
+{
+  if (email.empty())
+    return true;
+
+  auto const atPos = find(begin(email), end(email), '@');
+  if (atPos == end(email))
+    return false;
+
+  // There should be only one '@' sign.
```

Comment:
http://stackoverflow.com/questions/12355858/how-many-symbol-can-be-in-an-email-address